src/java.net.http/share/classes/jdk/internal/net/http/LimitingSubscriber.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (c) 2025, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package jdk.internal.net.http;
+
+import jdk.internal.net.http.ResponseSubscribers.TrustedSubscriber;
+import jdk.internal.net.http.common.Utils;
+
+import java.io.IOException;
+import java.net.http.HttpResponse.BodySubscriber;
+import java.nio.Buffer;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.Flow.Subscription;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A subscriber limiting the maximum number of bytes that are allowed to be consumed by a downstream subscriber.
+ *
+ * @param <T> the response type
+ */
+public final class LimitingSubscriber<T> implements TrustedSubscriber<T> {
+
+    private final BodySubscriber<T> downstreamSubscriber;
+
+    private final long capacity;
+
+    private State state;
+
+    private long length;
+
+    private interface State {
+
+        enum Terminated implements State { INSTANCE }
```

Comment:
This was a bit confusing to read. I find the following to be a bit more easier to read:

```java
	private interface State {
		State TERMINATED = new State() {};

		record Subscribed(Object subscription) implements State {}
	}
```
The call sites would then just use `State.TERMINATED`. 

However, if you prefer it in its current form, it's fine with me.

Reply:
Fixed in 23f79a20476f784058801454a28d37f37df92982.